examples/defaulter-gen/generators/defaulter.go

```diff
@@ -899,80 +906,89 @@ func (n *callNode) writeCalls(varName string, isVarPointer bool, sw *generator.S
 	}
 }
 
+func getTypeZeroValue(t string) (interface{}, error) {
+	defaultZero, ok := typeZeroValue[t]
+	if !ok {
+		return nil, fmt.Errorf("Cannot find zero value for type %v in typeZeroValue", t)
+	}
+
+	// To generate the code for empty string, they must be quoted
+	if defaultZero == "" {
+		defaultZero = strconv.Quote(defaultZero.(string))
+	}
+	return defaultZero, nil
+}
+
 func (n *callNode) writeDefaulter(varName string, index string, isVarPointer bool, sw *generator.SnippetWriter) {
 	if n.defaultValue == "" {
 		return
 	}
-	varPointer := varName
-	if !isVarPointer {
-		varPointer = "&" + varPointer
-	}
-
 	args := generator.Args{
 		"defaultValue": n.defaultValue,
-		"varPointer":   varPointer,
 		"varName":      varName,
 		"index":        index,
 		"varDepth":     n.defaultDepth,
 		"varType":      n.defaultType,
 	}
 
-	if n.index {
-		sw.Do("if reflect.ValueOf($.var$[$.index$]).IsZero() {\n", generator.Args{"var": varName, "index": index})
-		if n.defaultIsPrimitive {
-			if n.defaultDepth > 0 {
-				sw.Do("var ptrVar$.varDepth$ $.varType$ = $.defaultValue$\n", args)
-				for i := n.defaultDepth; i > 0; i-- {
-					sw.Do("ptrVar$.ptri$ := &ptrVar$.i$\n", generator.Args{"i": fmt.Sprintf("%d", i), "ptri": fmt.Sprintf("%d", (i - 1))})
-				}
-				sw.Do("$.varName$[$.index$] = ptrVar0", args)
-			} else {
-				sw.Do("$.varName$[$.index$] = $.defaultValue$", args)
-			}
-		} else {
-			sw.Do("if err := json.Unmarshal([]byte(`$.defaultValue$`), $.varPointer$[$.index$]); err != nil {\n", args)
-			sw.Do("panic(err)\n", nil)
-			sw.Do("}\n", nil)
+	if n.defaultIsPrimitive {
+		defaultZero, err := getTypeZeroValue(n.defaultType)
+		if err != nil {
+			klog.Error(err)
 		}
+		args["defaultZero"] = defaultZero
+	}
+
+	variablePlaceholder := ""
+
+	if n.index {
+		// Defaulting for array
+		variablePlaceholder = "$.varName$[$.index$]"
 	} else if n.key {
-		mapDefaultVar := index + "_default"
+		// Defaulting for map
+		variablePlaceholder = "$.varName$[$.index$]"
+		mapDefaultVar := args["index"].(string) + "_default"
 		args["mapDefaultVar"] = mapDefaultVar
-		sw.Do("if reflect.ValueOf($.var$[$.index$]).IsZero() {\n", generator.Args{"var": varName, "index": index})
+	} else {
+		// Defaulting for primitive type
+		variablePlaceholder = "$.varName$"
+	}
 
-		if n.defaultIsPrimitive {
-			if n.defaultDepth > 0 {
-				sw.Do("var ptrVar$.varDepth$ $.varType$ = $.defaultValue$\n", args)
-				for i := n.defaultDepth; i > 0; i-- {
-					sw.Do("ptrVar$.ptri$ := &ptrVar$.i$\n", generator.Args{"i": fmt.Sprintf("%d", i), "ptri": fmt.Sprintf("%d", (i - 1))})
-				}
-				sw.Do("$.varName$[$.index$] = ptrVar0", args)
-			} else {
-				sw.Do("$.varName$[$.index$] = $.defaultValue$", args)
+	variablePointer := variablePlaceholder
```

Comment:
This one is only used in one place below, can you reduce the scope of this variable? L986